File: python/best_time_to_buy_and_sell_stock_with_cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        sold, held, reset = float("-inf"), float("-inf"), 0
            for price in prices:
                preSold = sold
                sold = held + prices[i]
                held = max(held, reset)
                reset = max(preSold, reset)
            return max(sold, reset)

File: javascript/best_time_to_buy_and_sell_stock_with_cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let sold = -Infinity;
  let held = -Infinity;
  let reset = 0;
  prices.forEach(price => {
    const preSold = sold;
    sold = held + price;
    held = Math.max(held, reset - price);
    reset = Math.max(preSold, reset);
  });
  return Math.max(sold, reset);
};
